collections/sample.ts

```diff
@@ -25,12 +46,55 @@ function randomInteger(lower: number, upper: number): number {
  * import { assertArrayIncludes } from "@std/assert";
  *
  * const numbers = [1, 2, 3, 4];
- * const random = sample(numbers);
+ * const sampled = sample(numbers);
+ *
+ * assertArrayIncludes(numbers, [sampled]);
+ * ```
  *
- * assertArrayIncludes(numbers, [random]);
+ * @example Using `weights` option
+ * ```ts no-assert
+ * import { sample } from "@std/collections/sample";
+ *
+ * const weighted = new Map([["a", 5], ["b", 3], ["c", 2]]);
+ * const result = sample([...weighted.keys()], { weights: [...weighted.values()] });
+ * // gives "a" 50% of the time, "b" 30% of the time, and "c" 20% of the time
  * ```
  */
-export function sample<T>(array: readonly T[]): T | undefined {
+export function sample<T>(
+  array: readonly T[],
+  options?: Partial<SampleOptions>,
+): T | undefined {
+  const { random, weights } = { ...defaultOptions, ...options };
+
+  if (weights) {
+    assertEquals(
+      weights.length,
+      array.length,
+      "The length of the weights array must match the length of the input array",
+    );
```

Comment:
We now have policy of not using `@std/assert` for checking user's input. https://github.com/denoland/std/blob/4432ba1c1bc02abc30e2e2238bc50e0127000fb9/.github/CONTRIBUTING.md#assertions

Can you replace this with if-statement check and throw RangeError in this case?

Could you also do the same for the other usages of `@std/assert`?

Reply:
I'll remove those. What about `unreachable` though? The rationale from https://github.com/denoland/std/issues/4865 mentions three reasons not to use asserts in implementation code, of which two don't apply to `unreachable` (there should never be any need to handle `unreachable` errors, and an `unreachable` error fully captures its semantics as it's very literally an assertion). The third reason, introduction of a dependency, does still apply.